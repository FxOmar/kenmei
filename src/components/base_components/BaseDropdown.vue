<template lang="pug">
  .base-dropdown
    button(@click='open = !open')
      icon-dots-vertical.h-5.w-5
    button.fixed.inset-0.h-full.w-full.cursor-default.z-20(
      ref="clickOffButton"
      v-show='open'
      @click='open = false'
      tabindex='-1'
    )
    transition(name="dropdown")
      .dropdown-body(v-show='open' :class="sizes[size]")
        .rounded-md.bg-white.shadow-xs
          .py-1
            a.group(
              v-for="(item, index) in items"
              href='#'
              :key="index"
              :class="classes(item)"
              @click.prevent="handleClick(item)"
            )
              component.icon.group-hover_text-gray-500.group-focus_text-gray-500(
                v-if="item.icon"
                :is="item.icon"
              )
              | {{ item.text }}
</template>

<script>
  export default {
    props: {
      items: {
        type: Array,
        required: true,
      },
      size: {
        type: String,
        default: '4xl',
      },
    },
    data() {
      return {
        open: false,
        sizes: {
          lg: 'w-32',
          '3xl': 'w-48',
          '4xl': 'w-56',
        },
      };
    },
    created() {
      const handleEscape = (e) => {
        if (e.key === 'Esc' || e.key === 'Escape') { this.open = false; }
      };

      document.addEventListener('keydown', handleEscape);
      this.$once('hook:beforeDestroy', () => {
        document.removeEventListener('keydown', handleEscape);
      });
    },
    methods: {
      classes(item) {
        return {
          disabled: item.disabled,
        };
      },
      handleClick(item) {
        if (item.disabled) return;

        this.$emit('click', item.text);
        this.open = false;
      },
    },
  };
</script>

<style lang="scss" scoped>
  .base-dropdown {
    -webkit-tap-highlight-color: transparent;

    @apply relative inline-block text-left;
  }

  a {
    @apply flex items-center px-4 py-2 text-sm leading-5 text-gray-700;

    &:hover {
      @apply bg-gray-100 text-gray-900;
    }

    &:focus {
      @apply outline-none bg-gray-100 text-gray-900;
    }

    .disabled {
      @apply opacity-50 cursor-not-allowed;

      &:focus {
        @apply bg-white text-gray-700;
      }
    }
  }

  button {
    @apply flex items-center text-gray-400;

    &:hover {
      @apply text-gray-600;
    }

    &:focus {
      @apply outline-none text-gray-600;
    }
  }

  .icon {
    @apply mr-3 h-5 w-5 text-gray-400;
  }

  .dropdown-body {
    min-width: -webkit-max-content;
    min-width: -moz-max-content;
    min-width: max-content;
    @apply origin-top-right absolute right-0 mt-2 rounded-md shadow-lg;
    @apply z-20;
  }

  .dropdown-enter-active {
    @apply ease-out duration-100;
  }

  .dropdown-leave-active {
    @apply ease-in duration-75;
  }

  .dropdown-enter,
  .dropdown-leave-to {
    @apply transform opacity-0 scale-95;
  }

  .dropdown-enter-to,
  .dropdown-leave {
    @apply transform opacity-100 scale-100;
  }
</style>
