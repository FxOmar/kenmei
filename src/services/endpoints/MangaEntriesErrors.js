import { secure } from '@/modules/axios';

/* eslint-disable import/prefer-default-export */
export const postMangaEntriesErrors = (entriesIDs, issueID) => secure
  .post('/api/v1/manga_entries_errors/', { ids: entriesIDs, issue_id: issueID })
  .then(() => true)
  .catch(() => false);
/* eslint-enable import/prefer-default-export */
