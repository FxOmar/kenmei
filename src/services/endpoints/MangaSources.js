import { secure } from '@/modules/axios';

/* eslint-disable import/prefer-default-export */
export const getMangaSources = (mangaSeriesID) => secure
  .get('/api/v1/manga_sources', {
    params: { manga_source: { filter: { manga_series: mangaSeriesID } } },
  })
  .then((response) => response.data)
  .catch(() => false);
/* eslint-enable import/prefer-default-export */
