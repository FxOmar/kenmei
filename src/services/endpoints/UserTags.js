import { secure } from '@/modules/axios';

const baseURL = '/api/v1/user_tags';

export const index = () => secure
  .get(baseURL)
  .then((response) => response)
  .catch((request) => request.response);
export const create = (userTag) => secure
  .post(baseURL, { user_tag: userTag })
  .then((response) => response)
  .catch((request) => request.response);
export const update = (userTag) => secure
  .put(`${baseURL}/${userTag.id}`, { user_tag: userTag })
  .then((response) => response)
  .catch((request) => request.response);
export const destroy = (id) => secure
  .delete(`${baseURL}/${id}`)
  .then((response) => response)
  .catch((request) => request.response);
