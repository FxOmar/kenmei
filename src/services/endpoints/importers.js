import { secure } from '@/modules/axios';

/* eslint-disable import/prefer-default-export */
export const postTrackrMoe = filteredLists => secure
  .post('/api/v1/importers/trackr_moe', { lists: filteredLists })
  .then(() => true)
  .catch(() => false);
/* eslint-enable import/prefer-default-export */
