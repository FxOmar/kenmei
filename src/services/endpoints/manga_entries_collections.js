import { secure } from '@/modules/axios';

const baseURL = '/api/v1/manga_entries_collections';

/* eslint-disable import/prefer-default-export */
export const update = (attributes) => secure
  .post(baseURL, { manga_entries_collection: attributes })
  .then((response) => response)
  .catch((request) => request.response);
/* eslint-enable import/prefer-default-export */
