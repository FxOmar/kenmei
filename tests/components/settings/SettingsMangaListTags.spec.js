import { Message } from 'element-ui';
import flushPromises from 'flush-promises';
import MangaListTags from '@/components/settings/SettingsMangaListTags.vue';
import BaseDropdown from '@/components/base_components/BaseDropdown.vue';
import * as endpoint from '@/services/endpoints/UserTags';

const localVue = createLocalVue();

// To avoid missing directive Vue warnings
localVue.directive('loading', true);

describe('SettingsMangaListTags.vue', () => {
  afterEach(() => { jest.restoreAllMocks(); });

  describe('when mounted', () => {
    let tagsSpy;

    beforeEach(() => {
      tagsSpy = jest.spyOn(endpoint, 'index');
    });

    afterEach(() => {
      jest.restoreAllMocks();
    });

    describe('and API returns tags', () => {
      it('shows list of tags', async () => {
        tagsSpy.mockResolvedValue({
          status: 200,
          data: factories.userTag.buildList(1),
        });

        const mangaListTags = shallowMount(MangaListTags, { localVue });

        await flushPromises();

        expect(mangaListTags.find('ul').html()).toMatchSnapshot();
      });

      it('shows No tags added yet without tags', async () => {
        tagsSpy.mockResolvedValue({ status: 200, data: [] });

        const mangaListTags = shallowMount(MangaListTags, { localVue });

        await flushPromises();

        expect(mangaListTags.text()).toContain('No tags added yet');
      });
    });

    describe('and API returns an error', () => {
      it('shows Something went wrong message', async () => {
        const errorMessageMock = jest.spyOn(Message, 'error');
        tagsSpy.mockResolvedValue({ status: 500 });

        shallowMount(MangaListTags, { localVue });

        await flushPromises();

        expect(errorMessageMock).toHaveBeenCalledWith(
          'Something went wrong, try refreshing the page',
        );
      });
    });
  });

  describe('when selecting an action for a tag', () => {
    let mangaListTags;

    const tag1 = factories.userTag.build({ name: 'Tag 1' });
    const tag2 = factories.userTag.build({ name: 'Tag 2' });

    beforeEach(() => {
      mangaListTags = shallowMount(MangaListTags, {
        localVue,
        data() { return { tags: [tag1, tag2] }; },
      });
    });

    describe('and action is Edit', () => {
      it('shows modal with populated tag details', async () => {
        await mangaListTags.findAllComponents(BaseDropdown).at(1).vm.$emit(
          'click', 'Edit',
        );

        expect(mangaListTags.vm.form.id).toEqual(tag2.id);
        expect(mangaListTags.vm.form.name).toEqual(tag2.name);
        expect(mangaListTags.vm.form.description).toEqual(tag2.description);
        expect(mangaListTags.vm.modalVisible).toBeTruthy();
      });
    });

    describe('and action is Delete', () => {
      it('shows modal with populated tag details', async () => {
        await mangaListTags.findAllComponents(BaseDropdown).at(1).vm.$emit(
          'click', 'Delete',
        );

        expect(mangaListTags.vm.form.id).toEqual(tag2.id);
      });
    });
  });

  describe('when creating a tag', () => {
    let tagsSpy;
    let mangaListTags;

    const tag1 = factories.userTag.build({ name: 'Tag 1' });

    beforeEach(() => {
      tagsSpy = jest.spyOn(endpoint, 'create');

      mangaListTags = mount(MangaListTags, {
        localVue,
        data() { return { tags: [tag1] }; },
      });
    });

    it('shows tag creation modal', async () => {
      mangaListTags.find({ ref: 'openTagCreationModal' }).trigger('click');

      expect(mangaListTags.vm.modalVisible).toBeTruthy();
    });

    describe('and creation is successful', () => {
      it('adds new tag to the tags list', async () => {
        const newTag = factories.userTag.build({
          id: null, name: 'New Tag', description: 'Description',
        });

        tagsSpy.mockResolvedValue({ status: 200, data: newTag });

        await mangaListTags.find({ ref: 'openTagCreationModal' })
          .trigger('click');
        await mangaListTags.setData({ form: newTag });

        mangaListTags.find({ ref: 'addTagButton' }).trigger('click');

        await flushPromises();

        expect(tagsSpy).toHaveBeenCalledWith(newTag);
        expect(mangaListTags.vm.modalVisible).toBeFalsy();
        expect(mangaListTags.vm.modalLoading).toBeFalsy();
        expect(mangaListTags.vm.tags).toEqual([tag1, newTag]);
        expect(mangaListTags.vm.form).toEqual({
          id: null,
          name: '',
          description: '',
        });
      });
    });

    describe('and creation is unsuccessful', () => {
      it('shows validation errors', async () => {
        const errorMessageMock = jest.spyOn(Message, 'error');
        tagsSpy.mockResolvedValue({ status: 500, data: 'Name exists' });

        await mangaListTags.find({ ref: 'openTagCreationModal' })
          .trigger('click');
        await mangaListTags.setData({
          form: factories.userTag.build({ id: null }),
        });

        mangaListTags.find({ ref: 'addTagButton' }).trigger('click');

        await flushPromises();

        expect(mangaListTags.vm.modalVisible).toBeTruthy();
        expect(mangaListTags.vm.modalLoading).toBeFalsy();
        expect(errorMessageMock).toHaveBeenCalledWith('Name exists');
      });
    });
  });

  describe('when updating a tag', () => {
    let tagsSpy;
    let mangaListTags;

    const tag1 = factories.userTag.build({ name: 'Tag 1' });
    const tag2 = factories.userTag.build({ name: 'Tag 2' });

    beforeEach(() => {
      tagsSpy = jest.spyOn(endpoint, 'update');

      mangaListTags = mount(MangaListTags, {
        localVue,
        data() { return { tags: [tag1, tag2], modalVisible: true }; },
      });
    });

    afterEach(() => {
      jest.restoreAllMocks();
    });

    describe('and update is successful', () => {
      it('updates tag data', async () => {
        const updatedTag = factories.userTag.build({
          id: tag1.id, name: 'New Tag', description: '',
        });

        tagsSpy.mockResolvedValue({ status: 200, data: updatedTag });

        await mangaListTags.setData({ form: updatedTag });

        mangaListTags.find({ ref: 'addTagButton' }).trigger('click');

        await flushPromises();

        expect(mangaListTags.vm.modalVisible).toBeFalsy();
        expect(mangaListTags.vm.modalLoading).toBeFalsy();
        expect(mangaListTags.vm.tags).toEqual([
          updatedTag,
          tag2,
        ]);
        expect(mangaListTags.vm.form).toEqual({
          id: null,
          name: '',
          description: '',
        });
      });
    });

    describe('and update is unsuccessful', () => {
      it('shows validation errors', async () => {
        const errorMessageMock = jest.spyOn(Message, 'error');
        tagsSpy.mockResolvedValue({ status: 500, data: 'Name exists' });

        await mangaListTags.setData({ form: factories.userTag.build() });

        mangaListTags.find({ ref: 'addTagButton' }).trigger('click');

        await flushPromises();

        expect(mangaListTags.vm.modalVisible).toBeTruthy();
        expect(mangaListTags.vm.modalLoading).toBeFalsy();
        expect(errorMessageMock).toHaveBeenCalledWith('Name exists');
      });
    });
  });

  describe('when deleting a tag', () => {
    let tagsSpy;
    let mangaListTags;

    const tag1 = factories.userTag.build();
    const tag2 = factories.userTag.build();

    beforeEach(() => {
      tagsSpy = jest.spyOn(endpoint, 'destroy');

      mangaListTags = mount(MangaListTags, {
        localVue,
        data() { return { tags: [tag1, tag2] }; },
      });
    });

    afterEach(() => {
      expect(tagsSpy).toHaveBeenCalledWith(tag2.id);
      expect(mangaListTags.vm.form).toEqual({
        id: null,
        name: '',
        description: '',
      });
    });

    describe('and deletion is successful', () => {
      it('removes tag from tags list', async () => {
        tagsSpy.mockResolvedValue({ status: 200 });

        await mangaListTags.findAllComponents(BaseDropdown).at(1).vm.$emit(
          'click', 'Delete',
        );

        await flushPromises();

        expect(mangaListTags.vm.tags).toEqual([tag1]);
      });
    });

    describe('and deletion is unsuccessful', () => {
      it('shows validation errors', async () => {
        const errorMessageMock = jest.spyOn(Message, 'error');

        tagsSpy.mockResolvedValue({ status: 500, data: 'Not your tag' });

        await mangaListTags.findAllComponents(BaseDropdown).at(1).vm.$emit(
          'click', 'Delete',
        );

        await flushPromises();

        expect(errorMessageMock).toHaveBeenCalledWith('Not your tag');
      });
    });
  });
});
