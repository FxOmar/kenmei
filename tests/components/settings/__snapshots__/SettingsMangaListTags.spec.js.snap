// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsMangaListTags.vue when mounted and API returns tags shows list of tags 1`] = `
<ul>
  <li class="flex items-center px-4 py-4 sm_px-6">
    <div class="min-w-0 flex-1 flex items-center">
      <div class="min-w-0 flex-1 items-center grid grid-col-1 space-y-4 sm_space-y-0 sm_grid-cols-4 sm_gap-4">
        <div class="col-span-1">
          <base-badge-stub text="Isekai" type="primary" class="float-left"></base-badge-stub>
        </div>
        <div class="col-span-2 text-sm leading-5 text-gray-500"><span>MC ends up in another world</span></div>
        <div class="col-span-1 text-sm leading-5 text-gray-500"><span class="font-bold">1</span> manga entries</div>
      </div>
    </div>
    <base-dropdown-stub items="[object Object],[object Object]" size="lg"></base-dropdown-stub>
  </li>
</ul>
`;
